File: vstest/week2/p1012-connected-component-dfs/p1012.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>   // ✅ 여기 추가!
#include <deque>

#define DEBUG 0
using namespace std;

int dy[] = {-1, 0, 1, 0};
int dx[] = { 0, 1, 0,-1};

void display(int mat[][100], int n, int m) {
    cout << "\n";
    for (int y=0; y<n; y++){
        for (int x=0; x<m; x++){
            cout << mat[y][x] << " ";
        }
        cout << endl;
    }
    cout << "\n";
}

void dfs(int y, int x, int mat[][100], int N, int M){
    mat[y][x] = 2;
    if (DEBUG) display(mat, N, M);
    for (int i=0; i<4; i++){
        int ny = y + dy[i];
        int nx = x + dx[i];
        int is_land = 0;
        if (N > ny && ny >=0 && M > nx && nx >= 0){
            is_land = mat[ny][nx];
        }
        if (is_land == 1){
            dfs(ny, nx, mat, N, M);
        }
    }
}

int main(){
    ios::sync_with_stdio(false); // C I/O와 C++ I/O를 분리해서 더 빠르게
    cin.tie(NULL);
    cout.tie(NULL);

    int T, N, M, K;
    int Arr[100][100] = {};
    cin >> T;
    
    for (int t=1; t <= T; t++){
        cin >> M >> N >> K;
        fill(&Arr[0][0], &Arr[0][0] + 100*100, 0);
        
        for (int k=1; k<=K; k++){
            int y, x;
            cin >> x >> y;
            Arr[y][x] = 1;
        }
        if (DEBUG) display(Arr, N, M);

        int cnt = 0;
        for (int y=0; y<N; y++){
            for (int x=0; x<M; x++){
                if (Arr[y][x] == 1) {
                    dfs(y, x, Arr, N, M);
                    cnt += 1;
                }
            }
        }
        cout << cnt << "\n";
    }

    return 0;
}

File: vstest/week2/2B-connected-component-dfs/p1012.py
DEBUG = 0
# T = 1
# M, N, K = 10, 8, 17
# Pos=[ 
#     [0, 0], [1, 0], [1, 1], [4, 2], [4, 3], 
#     [4, 5], [2, 4], [3, 4], [7, 4], [8, 4], 
#     [9, 4], [7, 5], [8, 5], [9, 5], [7, 6], 
#     [8, 6], [9, 6],
# ]
# 1111
# 1010
# 1100
# 1000

# [0,0],[0,1],[0,2],[0,3]
# [1,0],[1,1],[1,2],[1,3]
# [2,0],[2,1],[2,2],[2,3]
# [3,0],[3,1],[3,2],[3,3]

def display(mat, n, m):
    print(">")
    for y in range(0,n):
        print(f"{mat[y]}")
        # for x in range(0,m):
        #     print(f"{map[y][x]} ")
    print("<")

def dfs(y,x, mat):
    mat[y][x] = 2
    if DEBUG: display(mat, N, M)
    display(MAT, N, M)
    for i in [0, 1, 2, 3]:
        ny = y + dy[i]
        nx = x + dx[i]
        is_land = 0
        if N > ny and ny >= 0 and M > nx and nx >= 0:
            is_land = mat[ny][nx]
        if is_land == 1:
            dfs(ny,nx, mat)

T = int(input())
for t in range(T):
    M, N, K = [int(el) for el in input().split()]
    Pos = [[int(el) for el in input().split()] for _ in range(K)]

    MAT = [[0] * M for _ in range(N)]

    for el in Pos:
        MAT[el[1]][el[0]] = 1

    if DEBUG: display(MAT, N, M)

    dy = [-1, 0, 1, 0]
    dx = [ 0, 1, 0,-1]

    cnt = 0
    for y in range(N):
        for x in range(M):
            if MAT[y][x] == 1:
                dfs(y,x, MAT)
                cnt+=1

    print(cnt, y, x)
display(MAT, N, M)